### PR TITLE
Data-plan system api-gateway

### PR DESCRIPTION
**Pull request by Brackleycassinga**
## Purpose changes

Issue Ref: #361 

**Data-plan api-gateway**

Endpoints 

- [x] Upload baseRates
- [x] Get baseRates
- [x] Get baseRate
- [x] Add package
- [x] Update package
- [x] Delete package

### CODE REVIEW

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+		logrus.Error(err)
+	}
+}
+
+func (r *Router) init() {
+	r.f = rest.NewFizzRouter(r.config.serverConf, pkg.SystemName, version.Version, r.config.debugMode)
+	v1 := r.f.Group("/v1", "Data-plan system ", "Data-plan  system version v1")
+
+	baseRates := v1.Group("/baseRates", "BaseRates", "BaseRates operations")
+	baseRates.POST("", formatDoc("Upload baseRates", ""), tonic.Handler(r.uploadBaseRateHandler, http.StatusOK))
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:95]: POST handler should return StatusCreated

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	return resp, nil
+}
+func (p *Router) getBaseRatesHandler(c *gin.Context, req *GetBaseRatesRequest) (*pbBaseRate.GetBaseRatesResponse, error) {
+	provider := c.Query("Provider")
+	effectiveAt := c.Query("EffectiveAt")
+	simType := c.Query("SimType")
+	country := c.Query("country") // shortcut for c.Request.URL.Query().Get("lastname")
+	to, err := strconv.ParseUint(c.Param("To"), 10, 64)
+	if err != nil {
+		logrus.Error(err)
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:151]: Again, logging bad inputs seems not enough to me.

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+		return nil, err
+	}
+
+	return resp, nil
+}
+func (p *Router) deletePackageHandler(c *gin.Context, req *DeletePackageRequest) (*pb.DeletePackageResponse, error) {
+	_id, err := strconv.ParseUint(c.Param("package"), 10, 64)
+	if err != nil {
+		logrus.Error(err)
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:194]: Bad input

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+		SmsVolume:   req.SmsVolume,
+		DataVolume:  req.DataVolume,
+		VoiceVolume: req.VoiceVolume,
+		OrgRatesId:  req.OrgRatesId,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+	res, err := p.clients.d.GetPackage(&pb.GetPackagesRequest{
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:222]: We should not be trying to get  back the item we've just updated.

**Comment by delvatt** [comment]
Hey @Brackleycassinga, Good job for your very first API Gateway in Go!  There is a couple of things that we need to see together, but we are on a good track!

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	return []fizz.OperationOption{func(info *openapi.OperationInfo) {
+		info.Summary = summary
+		info.Description = description
+	}}
+}
+func (p *Router) getBaseRateHandler(c *gin.Context, req *GetBaseRateRequest) (*pbBaseRate.GetBaseRateResponse, error) {
+
+	RateId, err := strconv.ParseUint(c.Param("baseRate"), 10, 64)
+	if err != nil {
+		logrus.Error(err)
```
[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/data-plan/api-gateway/pkg/rest/router.go:117]: Fixed.I have Return the error in request
```
 return nil, &rest.HttpError{HttpCode: http.StatusBadRequest,
			Message: "rateId is required"}
```

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+	baseRates := v1.Group("/baseRates", "BaseRates", "BaseRates operations")
+	baseRates.POST("", formatDoc("Upload baseRates", ""), tonic.Handler(r.uploadBaseRateHandler, http.StatusCreated))
+	baseRates.GET("/:baseRate", formatDoc("Get BaseRate", ""), tonic.Handler(r.getBaseRateHandler, http.StatusOK))
+	baseRates.GET("", formatDoc("Get BaseRates by country", ""), tonic.Handler(r.getBaseRatesHandler, http.StatusOK))
+
+	packages := v1.Group("/packages", "Packages", "Packages operations")
+	packages.PUT("", formatDoc("Add Package", ""), tonic.Handler(r.AddPackageHandler, http.StatusCreated))
+	packages.GET("/:package", formatDoc("Get package", ""), tonic.Handler(r.getPackageHandler, http.StatusOK))
+	packages.PATCH("", formatDoc("Update Package", ""), tonic.Handler(r.UpdatePackageHandler, http.StatusOK))
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:102]: PATCH is used to "update-by-patch" an existing resource. So it should be used over an identifiable REST resource, clearly specified as a path parameter.

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+	r.f = rest.NewFizzRouter(r.config.serverConf, pkg.SystemName, version.Version, r.config.debugMode)
+	v1 := r.f.Group("/v1", "Data-plan system ", "Data-plan  system version v1")
+
+	baseRates := v1.Group("/baseRates", "BaseRates", "BaseRates operations")
+	baseRates.POST("", formatDoc("Upload baseRates", ""), tonic.Handler(r.uploadBaseRateHandler, http.StatusCreated))
+	baseRates.GET("/:baseRate", formatDoc("Get BaseRate", ""), tonic.Handler(r.getBaseRateHandler, http.StatusOK))
+	baseRates.GET("", formatDoc("Get BaseRates by country", ""), tonic.Handler(r.getBaseRatesHandler, http.StatusOK))
+
+	packages := v1.Group("/packages", "Packages", "Packages operations")
+	packages.PUT("", formatDoc("Add Package", ""), tonic.Handler(r.AddPackageHandler, http.StatusCreated))
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:100]: Are you using PUT for creating a new resource? Why not using POST then ?

File: systems/data-plan/api-gateway/pkg/rest/router.go (hunk 2 above)
Comment (delvatt) [reply systems/data-plan/api-gateway/pkg/rest/router.go:159]: This is not resolved.

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+		return nil, &rest.HttpError{HttpCode: http.StatusBadRequest,
+			Message: "country is a mandatory query parameter"}
+	}
+	to, err := strconv.ParseUint(c.Param("To"), 10, 64)
+	if err != nil {
+		logrus.Error(err)
+	}
+	from, err := strconv.ParseUint(c.Param("From"), 10, 64)
+	if err != nil {
+		logrus.Error(err)
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:163]: Again ...

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+	}
+	resp, err := p.clients.d.GetBaseRate(&pbBaseRate.GetBaseRateRequest{
+		RateId: reqBody.RateId,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	return resp, nil
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:128]: What are we using that ` req *GetBaseRateRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+		FileURL:     reqBody.FileURL,
+		EffectiveAt: reqBody.EffectiveAt,
+		SimType:     pbBaseRate.SimType(pbBaseRate.SimType_value[reqBody.SimType]),
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	return resp, nil
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:146]: What are we using that ` req *UploadBaseRatesRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+		From:        from,
+		SimType:     pbBaseRate.SimType(pbBaseRate.SimType_value[simType]),
+		EffectiveAt: effectiveAt,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	return resp, nil
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:179]: What are we using that ` req *GetBaseRatesRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+	resp, err := p.clients.d.GetPackage(&pb.GetPackagesRequest{
+		Id:    reqBody.Id,
+		OrgId: ORG_ID,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	return resp, nil
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:196]: What are we using that ` req *GetPackagesRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+	}
+	resp, err := p.clients.d.DeletePackage(&pb.DeletePackageRequest{
+		Id:    reqBody.Id,
+		OrgId: ORG_ID,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+	return resp, nil
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:212]: What are we using that ` req *DeletePackageRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+		VoiceVolume: reqBody.VoiceVolume,
+		OrgRatesId:  reqBody.OrgRatesId,
+	})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+	
+	return resp , nil
+
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:237]: What are we using that ` req *UpdatePackageRequest` parameter for in the end?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+		OrgId:       reqBody.OrgId,
+		Duration:    reqBody.Duration,
+		OrgRatesId:  reqBody.OrgRatesId,
+		VoiceVolume: reqBody.VoiceVolume,
+		Active:      reqBody.Active,
+		DataVolume:  reqBody.DataVolume,
+		SmsVolume:   reqBody.SmsVolume,
+		SimType:     pb.SimType(pb.SimType_value[reqBody.SimType]),
+	})
+}
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:256]: What are we using that ` req *AddPackageRequest` parameter for in the end?

File: systems/data-plan/base-rate/pkg/db/rate_repo_test.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+		r := NewBaseRateRepo(&UkamaDbMock{
+			GormDb: gdb,
+		})
+
+		err = r.UploadBaseRates(rates)
+		assert.NoError(t, err)
+		err = mock.ExpectationsWereMet()
+		assert.NoError(t, err)
+	})
+}
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/db/rate_repo_test.go:213]: Good. But you are only testing canonical/happy paths.

File: systems/data-plan/base-rate/pkg/utils/utils.go [hunk 17]
```diff
@@ -32,17 +33,19 @@ func FetchData(url string) ([]RawRates, error) {
...
 
-	var rawRates []RawRates
-	if err := csvutil.Unmarshal(content, &rawRates); err != nil {
-		return nil, err
+	var r []RawRates
+	errorStr := "invalid CSV file data"
+	csvutil.Unmarshal(content, &r)
+
+	if len(r) == 0 || validations.IsEmpty(r[0].Country) ||
+		validations.IsEmpty(r[0].Network) ||
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/utils/utils.go:43]: Is an empty CSV file the same as an invalid CSV file in your domain?

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+		Active:      req.Active,
+		Duration:    req.Duration,
+		SmsVolume:   req.SmsVolume,
+		DataVolume:  req.DataVolume,
+		VoiceVolume: req.VoiceVolume,
+		OrgRatesId:  req.OrgRatesId,
+	})
+	if err != nil {
+		logrus.Error(err)
+		c.JSON(http.StatusBadRequest, gin.H{"error": err})
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:238]: @Brackleycassinga  Why is a faillure to update a package considered as a bad request?

File: systems/data-plan/README.md [hunk 19]
```diff
@@ -47,4 +47,159 @@ Data Plan system manage all the sim packages and base rates functionality. Data
 
 [Base rate sub-system](https://github.com/ukama/ukama/tree/main/systems/data-plan/base-rate)
 
-[Package sub-system](https://github.com/ukama/ukama/tree/main/systems/data-plan/packge)
\ No newline at end of file
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (delvatt) [reply systems/data-plan/README.md:60]: There's stil. that merge conflict inside that README file.

**Review by delvatt — COMMENTED** [review]
Hey @Brackleycassinga ,  almost there. Few more comments to go!

File: systems/data-plan/api-gateway/pkg/rest/router.go [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+		Active:      req.Active,
+		Duration:    req.Duration,
+		SmsVolume:   req.SmsVolume,
+		DataVolume:  req.DataVolume,
+		VoiceVolume: req.VoiceVolume,
+		OrgRatesId:  req.OrgRatesId,
+	})
+	if err != nil {
+		logrus.Error(err)
+		c.JSON(http.StatusNotFound, gin.H{"error": err})
```
Comment (delvatt) [inline systems/data-plan/api-gateway/pkg/rest/router.go:255]: @Brackleycassinga   if you fail to update a package, then this is an internal server error.

**Review by delvatt — COMMENTED** [review]
@Brackleycassinga  one last comment to go!

**Review by delvatt — COMMENTED** [review]
LGTM! @Brackleycassinga  feel free to move forward!

**Review by delvatt — APPROVED** [review]

**Pull request merged**